### PR TITLE
Legalize BFloat16 in NNC.

### PR DESCRIPTION
**Pull request by EikanWang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #85140
* #84402
* #84041
* __->__ #83988

Regarding BF16 support in NNC, we always convert the BF16 to FP32 and then compute with FP32. After the FP32 computation, we convert the FP32 result to BF16. This logic has been supported in [half_support.h](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/tensorexpr/half_support.h).

But the B16/FP32 conversion has not been supported by LLVM. Currently, LLVM only supports the BF16 in its front end but still cannot generate the assembly code. So we implement this feature in [llvm_codegen](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/tensorexpr/llvm_codegen.cpp) like aten implementation.

In this PR, it contains three points - Take BF16 as uint16, Convert BF16 to FP32 and Convert FP32 to BF16.

- Take BF16 as uint16 - [PR Change](https://github.com/pytorch/pytorch/pull/83988/files#diff-9d09ca2fce1c689ab43cd71795ab9b8b63447de950cf98ae8a18114e18d87e79R544-R546)
  We cannot naively convert map the BF16 to LLVM BF16 as the LLVM backend still does not support this data type as I mentioned. Meanwhile, the BF16 in PyTorch is a [structure](https://github.com/pytorch/pytorch/blob/master/c10/util/BFloat16.h#L73). Its realdata is uint16. So we also bitcast the BF16 tensor to uint16

- BF16 to FP32 conversion - [PR Change](https://github.com/pytorch/pytorch/pull/83988/files#diff-9d09ca2fce1c689ab43cd71795ab9b8b63447de950cf98ae8a18114e18d87e79R1057-R1061)
   we just need to shift the BF16 value left by 16bits and then bit cast the shifted value to FP32 

- FP32 to BF16 conversion - [PR Change](https://github.com/pytorch/pytorch/pull/83988/files#diff-9d09ca2fce1c689ab43cd71795ab9b8b63447de950cf98ae8a18114e18d87e79R1066-R1084)
   The conversion is kind of complex. Because we use RNE to implement it. The RNR to convert the FP32 to BF16 is as follows.
   ```C++
   uint16_t round_to_nearest_even(float src) {
     if (std::isnan(src)) {
       return UINT16_C(0x7FC0);
     } else {
       union {
         uint32_t U32;
         float F32;
       };
   
       F32 = src;
       uint32_t rounding_bias = ((U32 >> 16) & 1) + UINT32_C(0x7FFF);
       return static_cast<uint16_t>((U32 + rounding_bias) >> 16);
     }
   }
   ```

### CODE REVIEW

File: torch/csrc/jit/tensorexpr/llvm_codegen.cpp [hunk 1]
```diff
@@ -1033,6 +1053,38 @@ void LLVMCodeGenImpl::visit(CastPtr v) {
...
 
   // Scalar casts
+  if (is_from_bf16) {
+    auto lans = v->dtype().lanes();
+    value_ = irb_.CreateZExt(value_, llvmTypeToVec(IntTy_, lans));
+    auto vec_shl_val = toVec(llvm::ConstantInt::get(IntTy_, 16), lans);
+    value_ = irb_.CreateShl(value_, vec_shl_val);
+    value_ = irb_.CreateBitOrPointerCast(value_, llvmTypeToVec(FloatTy_, lans));
+    return;
+  }
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/llvm_codegen.cpp:1063]: It looks like the code only supports casts from BF16 to F32 and back. However, TE IR allows expressing casts from any dtype to any, e.g one could have cast BF16 to Int64. Could you please add the corresponding code or put an "non-implemented" assertion for it? Also, please add a comment explaining what's going on here (the description you have in the commit message would be good, for instance).

File: torch/csrc/jit/tensorexpr/llvm_codegen.cpp [hunk 2]
```diff
@@ -1033,6 +1053,38 @@ void LLVMCodeGenImpl::visit(CastPtr v) {
...
+
+  if (is_to_bf16) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/llvm_codegen.cpp:1065]: Ditto.

File: torch/csrc/jit/tensorexpr/llvm_codegen.cpp (hunk 1 above)
Comment (EikanWang) [reply torch/csrc/jit/tensorexpr/llvm_codegen.cpp:1063]: Got it, thanks for your comments. I will fix it by the end of today.

Comment (EikanWang) [reply torch/csrc/jit/tensorexpr/llvm_codegen.cpp]: Done

File: torch/csrc/jit/tensorexpr/llvm_codegen.cpp (hunk 2 above)
Comment (EikanWang) [reply torch/csrc/jit/tensorexpr/llvm_codegen.cpp:1074]: Done

File: torch/csrc/jit/tensorexpr/llvm_codegen.cpp [hunk 3]
```diff
@@ -1033,6 +1059,56 @@ void LLVMCodeGenImpl::visit(CastPtr v) {
...
+    //   STEP1: U32_VAL = BITCAST(F32_VAL)
+    //   STEP2: U32_VAL_TMP = U32_VAL >> 16
+    //   STEP3: U32_VAL_TMP = U32_VAL_TMP & 1
+    //   STEP4: ROUNDING_BIAS = U32_VAL_TMP + UINT32(0x7FFF)
+    //   STEP5: U32_VAL_TMP = U32_VAL + ROUNDING_BIAS
+    //   STEP6: BF16_VAL = static_cast<UINT16>(U32_VAL_TMP >> 16)
+    auto lans = v->src_value()->dtype().lanes();
+    auto shift_len = llvm::ConstantInt::get(IntTy_, 16);
+    auto one = llvm::ConstantInt::get(ShortTy_, 1);
+    auto rounding_bias = llvm::ConstantInt::get(ShortTy_, 32767);
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/llvm_codegen.cpp:1086]: Nit: s/32767/0x7FFF/ for consistency with the rest of the code.

**Review by ZolotukhinM — APPROVED** [review]
LGTM!

File: torch/csrc/jit/tensorexpr/llvm_codegen.cpp (hunk 3 above)
Comment (EikanWang) [reply torch/csrc/jit/tensorexpr/llvm_codegen.cpp]: Done

**Review by frank-wei — APPROVED** [review]

**Comment by EikanWang** [comment]
@frank-wei , this PR needs to combine #84041 as a single PR to pass the CI. Meanwhile, this PR is reverted, then I will apply this PR to #84041 . Are you okay with that?

**Pull request closed without merging**